First Draft: Cookies with Nextjs (#33384)
vercel/next.js: docs/guides/next-cookies.md

![nextjs-cookies](/docs/guides/images/nextjs-cookies.png)

Setting and getting cookies on the client-side with Next.js is exactly the same as you would do with any other web application. However, on the server-side, Next.js helps you one step further to set and get cookies (more on this later).

## Client-side Cookies in Next.js

You set cookies via,

```js
document.cookie = 'name=value'
```

You get cookies via,

```js
document.cookie
```

Reading `document.cookie` shows all the available cookies in the current context, where a semicolon separates each cookie.

The `document.cookie` property takes up different attributes. E.g. `max-age` lets you define when the cookie should expire, `secure` allows cookies to transmit only over a secure protocol such as HTTPS.

You can read more details about the `document.cookie` property on the [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/API/Document/cookie).

### Set and Get Cookies on the Server-side on the Client-side

Let's build a demo application to set a `theme` cookie to save the `light` or `dark` mode of your website.

![client-side-cookies](/docs/guides/images/client-cookies.png)

> Let's create a Next.js page that shows the current value of the cookie`theme` and allows you to change it to `light` or `dark`.

Reviewer: ```suggestion
First create a Next.js page that shows the current value of the cookie`theme` and allows you to change it to `light` or `dark`.
```

> ```js

Reviewer: This codeblock has the `changeTheme` function, without an explanation of what it is. Maybe we could just use the second codeblock with the full code instead?

export default function ClientSideCookiePage() {
  return (
    <div>
      <h1>Client-side Cookies in Next.js</h1>
      <div>THEME = {theme} </div>
      <button onClick={() => changeTheme('dark')}>Dark</button>
      <button onClick={() => changeTheme('light')}>Light</button>
    </div>
  )
}
```

You can use two React.js hooks to make this app easier to work with:

- [`useState`](https://reactjs.org/docs/hooks-state.html) hook to save the `theme` cookie value
- [`useEffect`](https://reactjs.org/docs/hooks-effect.html) hook to access the `document`, which is a side-effect

To understand how React hooks work, [check out the React documentation](https://reactjs.org/docs/hooks-intro.html).

Here's the basic structure of our code so far:

> ```js

Reviewer: Same as above, i think we can merge these codeblocks with the final code

import { useEffect, useState } from 'react'

export default function ClientSideCookiePage() {
  const [theme, setTheme] = useState()

  useEffect(() => {
    // Get cookies & find the value of the cookie called `theme`.
    // Set default cookie to `light` if not already set.
    // Store the `theme` color in state.
  })

  // Set theme & cookie to user selected color.
  const changeTheme = (color) => {}

  return (
    <div>
      <h1>Client-side Cookies in Next.js</h1>
      <div>THEME = {theme} </div>
      <button onClick={() => changeTheme('dark')}>Dark</button>
      <button onClick={() => changeTheme('light')}>Light</button>
    </div>
  )
}
```

And here's the complete code where you get the `theme` cookie and allow the user to set it to `light` or `dark` mode.

```js
import { useEffect, useState } from 'react'

export default function ClientSideCookiePage() {
  const [theme, setTheme] = useState()

  useEffect(() => {
    // Get cookies & find the value of the cookie called `theme`.

>     const color = document.cookie

Reviewer 2: docs example must use `decodeURIComponent` on the cookie value before returning.

      ?.split(';')
      .find((x) => x.trim().startsWith('theme'))
      ?.split('=')[1]

    // Set default cookie to `light` if not already set.
    if (!color) {
      color = 'light'

>       document.cookie = `theme=${color}`

Reviewer 2: Similarly, we must call `encodeURIComponent` on cookie value before setting.

    }

    // Store the `theme` color in state.
    setTheme(color)
  })

  // Set theme & cookie to user selected color.
  const changeTheme = (color) => {
    setTheme(color)
    document.cookie = `theme=${color}`
  }

  return (
    <div>
      <h1>Client-side Cookies in Next.js</h1>
      <div>THEME = {theme} </div>
      <button onClick={() => changeTheme('dark')}>Dark</button>
      <button onClick={() => changeTheme('light')}>Light</button>
    </div>
  )
}
```

This code sets `light` mode as the default theme. Then it allows the user to click a button to change that to `dark` mode or back to `light` mode. The state is kept by the `useState` React hook, which updates the UI when the cookie value is changed.

### TODO: _ADD VIDEO SET CLIENT COOKIES_

## Server-side cookies with Next.js

> Working with cookies on the server-side with Next.js is just like you would expect. However, Next.js provides you with a `context.req.cookies` JavaScript object containing all the cookies.

Reviewer: ```suggestion
Next.js provides you with a `context.req.cookies` JavaScript object containing all the cookies.
```

> If you export an async function called [`getServerSideProps`](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering) from a page, Next.js will pre-render this page on each request using the data returned by `getServerSideProps`.

Reviewer: ```suggestion
If you export an async function called [`getServerSideProps`](/docs/basic-features/data-fetching/get-server-side-props) from a page, Next.js will pre-render this page on each request using the data returned by `getServerSideProps`.
```
